src/mondefense.cpp

```diff
@@ -1,38 +1,47 @@
+#include "mondefense.h"
 #include "monster.h"
+#include "creature.h"
+#include "damage.h"
 #include "game.h"
 #include "rng.h"
 #include "line.h"
 #include "bodypart.h"
-#include "speech.h"
 #include "messages.h"
-#include "mondefense.h"
-#include <algorithm>
 
-
-void mdefense::zapback(monster *m, Creature *source, const projectile *proj)
+void mdefense::zapback(monster *const m, Creature *const source, projectile const* const proj)
 {
-    player *foe = dynamic_cast< player* >( source );
-    monster *othermon = dynamic_cast< monster* >( source );
-    if( rl_dist( m->pos(), source->pos() ) > 1 ) {
-        return; // Out of range
+    if ((proj) ||                              // Not a melee attack
+        (rng(0, 100) > m->def_chance) ||       // Attacker lucked out
+        (rl_dist(m->pos(), source->pos()) > 1) // Out of range
+    ) {
+        return;
     }
-    if( proj != nullptr || rng(0, 100) > m->def_chance ) {
-        return; // Not a melee attack or attacker lucked out
+
+    auto const is_eligible_player = [](player const *const foe) {
```

Comment:
What's wrong with lambdas? You seem to have a chip on your shoulder regarding them; they are nothing special, nothing complicated, nothing to shirk or shun. Why is using lambdas to solve a problem any different than other constructs? What is that makes lambdas uniquely bad?

How would you propose, here and in general, to:
1) preserve constness where things *should* be semantically const.
2) keep variable visibility and lifetime as short as possible.
3) avoid creating deeply nested "arrow" code.
4) avoid the mental context switch of having to hunt down a tiny out-of-line function.
5) here, preserve the short-circuiting semantics of the checks.

The lambda-free version you merged fails 1, 2, and 5. And yes, *here* it "doesn't matter", but as a general principle, general ideal to aim for, I think that it does matter.